src/contracts/deposits-registry/deposits-registry.service.ts

```diff
@@ -0,0 +1,249 @@
+  }
+
+  public async initialize() {
+    await this.store.initialize();
+    const cachedEvents = await this.store.getEventsCache();
+    await this.sanityChecker.initialize(cachedEvents);
+
+    await this.updateEventsCache();
+  }
+
+  /**
+   * Gets node operators data from cache
+   * @returns event group
+   */
+  public async getCachedEvents(): Promise<VerifiedDepositEventsCache> {
+    const { headers, ...rest } = await this.store.getEventsCache();
+    const deploymentBlock = await this.fetcher.getDeploymentBlockByNetwork();
+
+    return {
+      headers: {
+        ...headers,
+        startBlock: Math.max(headers.startBlock, deploymentBlock),
+        endBlock: Math.max(headers.endBlock, deploymentBlock),
+      },
+      ...rest,
+    };
+  }
+
+  /**
+   * Updates the cache deposited events
+   * The last N blocks are not stored, in order to avoid storing reorganized blocks
+   */
+  public async updateEventsCache(): Promise<void> {
+    const fetchTimeStart = performance.now();
+
+    const [finalizedBlock, initialCache] = await Promise.all([
+      this.providerService.getBlock(this.finalizedTag),
+      this.getCachedEvents(),
+    ]);
+
+    const { number: finalizedBlockNumber, hash: finalizedBlockHash } =
+      finalizedBlock;
+    const firstNotCachedBlock = initialCache.headers.endBlock + 1;
+
+    const totalEventsCount = initialCache.data.length;
+    let newEventsCount = 0;
+
+    // verify blockchain
+    const isCacheValid = this.sanityChecker.verifyCacheBlock(
+      initialCache,
+      finalizedBlockNumber,
+    );
+
+    if (!isCacheValid) return;
+
+    let lastIndexedEvent: VerifiedDepositEvent | undefined = undefined;
+
+    for (
+      let block = firstNotCachedBlock;
+      block <= finalizedBlockNumber;
+      block += DEPOSIT_EVENTS_STEP
+    ) {
+      const chunkStartBlock = block;
+      const chunkToBlock = Math.min(
+        finalizedBlockNumber,
+        block + DEPOSIT_EVENTS_STEP - 1,
+      );
+
+      const chunkEventGroup = await this.fetcher.fetchEventsFallOver(
+        chunkStartBlock,
+        chunkToBlock,
+      );
+
+      await this.sanityChecker.verifyEventsChunk(
+        chunkStartBlock,
+        chunkToBlock,
+        chunkEventGroup.events,
+      );
+
+      // Even if the cache is not valid we can't help but write it down
+      // because the delay in updating the cache will eventually cause
+      // the getAllDepositedEvents method to take a very long time to process, as changes
+      // will be accumulated and not processed.
+      await this.store.insertEventsCacheBatch({
+        headers: {
+          ...initialCache.headers,
+          endBlock: chunkEventGroup.endBlock,
+        },
+        data: chunkEventGroup.events,
+      });
+
+      newEventsCount += chunkEventGroup.events.length;
+
+      const lastEventFromGroup =
+        chunkEventGroup.events[chunkEventGroup.events.length - 1];
+
+      if (lastEventFromGroup) lastIndexedEvent = lastEventFromGroup;
```

Comment:
do we need  `if (lastEventFromGroup)` ?

Reply:
Yes, we can have ‘empty’ gaps and because of that we may not record the last event.

Reply:
maybe we need comment it